jenkins-master/jobs/scripts/workspace/submission.py

```diff
@@ -0,0 +1,375 @@
+    parser.add_argument('--test-type',
+                        choices=config['test_types'].keys(),
+                        required=True,
+                        help='The type of test for building the job name and symbol.')
+    parser.add_argument('--repository',
+                        required=True,
+                        help='The repository name the build was created from.')
+    parser.add_argument('--revision',
+                        required=True,
+                        help='The revision of the build.')
+    parser.add_argument('--build-state',
+                        choices=BUILD_STATES,
+                        required=True,
+                        help='The state of the build')
+    parser.add_argument('venv_path',
+                        help='Path to the virtual environment to use.')
+
+    aws_group = parser.add_argument_group('AWS', 'Arguments for Amazon S3')
+    aws_group.add_argument('--aws-bucket',
+                           default=os.environ.get('AWS_BUCKET'),
+                           help='The S3 bucket name.')
+    aws_group.add_argument('--aws-key',
+                           default=os.environ.get('AWS_ACCESS_KEY_ID'),
+                           help='Access key for Amazon S3.')
+    aws_group.add_argument('--aws-secret',
+                           default=os.environ.get('AWS_SECRET_ACCESS_KEY'),
+                           help='Access secret for Amazon S3.')
+
+    treeherder_group = parser.add_argument_group('treeherder', 'Arguments for Treeherder')
+    treeherder_group.add_argument('--treeherder-url',
+                                  default=os.environ.get('TREEHERDER_URL'),
+                                  help='URL to the Treeherder server.')
+    treeherder_group.add_argument('--treeherder-client-id',
+                                  default=os.environ.get('TREEHERDER_CLIENT_ID'),
+                                  help='Client ID for submission to Treeherder.')
+    treeherder_group.add_argument('--treeherder-secret',
+                                  default=os.environ.get('TREEHERDER_SECRET'),
+                                  help='Secret for submission to Treeherder.')
+
+    update_group = parser.add_argument_group('update', 'Arguments for update tests')
+    update_group.add_argument('--update-number',
+                              action=JenkinsDefaultValueAction,
+                              help='The number of the partial update: today - N days')
+
+    return vars(parser.parse_args())
+
+
+if __name__ == '__main__':
+    kwargs = parse_args()
+
+    # Activate the environment, and create if necessary
+    import environment
+    if environment.exists(kwargs['venv_path']):
+        environment.activate(kwargs['venv_path'])
+    else:
+        environment.create(kwargs['venv_path'], os.path.join(here, 'requirements.txt'))
+
+    # Can only be imported after the environment has been activated
```

Comment:
Oh, I see. This script is running naked without a virtual env, and environment.create makes sure all of the requirements are there so that code following this will work.

The comment needs to make this clearer.